In [2]:
import numpy as np
import pandas as pd
import matplotlib as plt

In [3]:
df = pd.read_csv("../data/raw/17th.csv", sep=";")

# 17th Lok Sabha Analysis — Data Preparation

This notebook performs the initial data preparation and quality checks
for the 17th Lok Sabha dataset (2019–2024).

The purpose of this stage is to understand the structure and quality
of the dataset before performing exploratory data analysis (EDA)
and visualization.

The raw dataset is kept unchanged in the `data/raw/` directory.

In [4]:
df.head()

,Name,Constituency,Minister,Attendance,Debates,Questions,Private Member Bills,State,Party,Nature of membership,...,Debates (Statement By Minister),Debates (Clarification On The Statement By Minister),Debates (Felicitations),Questions (Consumer Affairs),Questions (Micro),Debates (Ls - Half-An-Hour Discussions (Rule-55)),Questions (Atomic Energy),Questions (Heavy Industries),Questions (Space),"Questions (Personnel, Public Grievances And Pensions)"
0,Abdul Khaleque,Barpeta,No,92.00%,68,279,8,Assam,Indian National Congress,Elected,...,0,0,0,0,0,0,0,0,0,0
1,Abhishek Banerjee,Diamond Harbour,No,19.80%,2,84,0,West Bengal,All India Trinamool Congress,Elected,...,0,0,0,0,0,0,0,0,0,0
2,Abu Hasem Khan Choudhury,Maldaha Dakshin,No,55.93%,5,12,0,West Bengal,Indian National Congress,Elected,...,0,0,0,0,0,0,0,0,0,0
3,Abu Taher Khan,Murshidabad,No,63.07%,6,4,0,West Bengal,All India Trinamool Congress,Elected,...,0,0,0,0,0,0,0,0,0,0
4,Achyutananda Samanta,Kandhamal,No,65.47%,33,169,0,Odisha,Biju Janata Dal,Elected,...,0,0,0,0,0,0,0,0,0,0


In [5]:
df.info

<bound method DataFrame.info of                          Name     Constituency Minister Attendance  Debates  \
0              Abdul Khaleque          Barpeta       No     92.00%       68   
1           Abhishek Banerjee  Diamond Harbour       No     19.80%        2   
2    Abu Hasem Khan Choudhury  Maldaha Dakshin       No     55.93%        5   
3              Abu Taher Khan      Murshidabad       No     63.07%        6   
4        Achyutananda Samanta        Kandhamal       No     65.47%       33   
..                        ...              ...      ...        ...      ...   
554        Vishnu Dutt Sharma        Khajuraho       No     56.53%       16   
555  Vivek Narayan Shejwalkar          Gwalior       No     96.73%       48   
556           V.K. Sreekandan         Palakkad       No     92.87%       63   
557            Y. Devendrappa          Bellary       No     83.27%       10   
558        Y.S. Avinash Reddy           Kadapa       No     41.47%        1   

     Questions  Pri

## Dataset Overview

The dataset contains information about representatives of the
17th Lok Sabha, including demographic information, party and
constituency details, attendance, debates, questions, and
Private Member Bills.

The dataset contains 559 representatives and 96 columns.

The data is derived from parliamentary activity information
and is used in this project for descriptive analysis of
parliamentary participation.

In [6]:
df.isna().sum()

Name                                                      0
Constituency                                              0
Minister                                                  0
Attendance                                               34
Debates                                                   0
                                                         ..
Debates (Ls - Half-An-Hour Discussions (Rule-55))         0
Questions (Atomic Energy)                                 0
Questions (Heavy Industries)                              0
Questions (Space)                                         0
Questions (Personnel, Public Grievances And Pensions)     0
Length: 96, dtype: int64

## Data Quality Finding

34 MPs have missing Attendance values.

The missing values are not random. Most correspond to MPs who were
ministers during the 17th Lok Sabha, while Om Birla is the Speaker.

Do not fill the missing Attendance values without a good reason.

For future analysis, Attendance should only be calculated among
MPs with reported attendance values.

In [7]:
df.duplicated().sum()

np.int64(0)

## Duplicate Check

No duplicate rows were found in the dataset.

Therefore, no rows were removed at this stage.

In [8]:
df['Minister'].value_counts()

Minister
No     488
Yes     71
Name: count, dtype: int64

In [9]:
df["Nature of membership"].value_counts()

Nature of membership
Elected    559
Name: count, dtype: int64

## Nature of Membership

All 559 representatives in the dataset are classified as `Elected`.

No missing values were found in this column.

In [10]:
df["End of Term"].value_counts()

End of Term
In Office     523
07-12-2023     11
22-03-2022      2
08-12-2023      2
13-12-2023      2
22-10-2021      1
28-02-2020      1
16-09-2020      1
30-05-2023      1
14-03-2022      1
29-03-2023      1
28-08-2020      1
15-12-2023      1
11-12-2023      1
22-02-2021      1
10-10-2022      1
02-03-2021      1
03-02-2021      1
21-07-2019      1
17-03-2021      1
18-05-2023      1
14-01-2023      1
14-09-2019      1
23-09-2020      1
Name: count, dtype: int64

## End of Term

- 559 representatives are included in the dataset.
- 523 were still in office at the end of the dataset period.
- 36 representatives have a recorded end date, meaning they left before the end of the term.
- These records should not be removed because they are valid members of the 17th Lok Sabha.

In [11]:
df["Gender"].value_counts()

Gender
Male      477
Female     82
Name: count, dtype: int64

## Gender Distribution

- Gender: 477 male and 82 female representatives.
- No missing Gender values were found.

The dataset is therefore strongly male-dominated.

In [12]:
df["Party"].value_counts()

Party
Bharatiya Janata Party                      308
Indian National Congress                     54
Dravida Munnetra Kazhagam                    24
All India Trinamool Congress                 23
Yuvajana Sramika Rythu Congress Party        23
Shiv Sena                                    19
Janata Dal (United)                          17
Biju Janata Dal                              12
Bahujan Samaj Party                          10
Bharat Rashtra Samithi                        9
Lok Jan Shakti Party                          7
Samajwadi Party                               6
Nationalist Congress Party                    5
Indian Union Muslim League                    4
Independent                                   4
Jammu and Kashmir National Conference         3
Shiromani Akali Dal                           3
Telugu Desam Party                            3
Communist Party of India (Marxist)            3
Apna Dal                                      2
All India Majlis-E-Ittehadul Musli

## Party Distribution

- The dataset contains 37 distinct party categories.
- Bharatiya Janata Party (BJP) has the largest representation with 308 members.
- Indian National Congress (INC) has 54 members.
- Several regional parties have significant representation.
- Many smaller parties have only 1–3 representatives.

Party names are retained as provided in the raw dataset.
Any grouping of parties will be performed later as part of the analysis,
rather than during initial data cleaning.

In [13]:
df['State'].value_counts()

State
Uttar Pradesh                  83
Maharashtra                    49
West Bengal                    43
Bihar                          42
Tamil Nadu                     40
Madhya Pradesh                 30
Karnataka                      29
Andhra Pradesh                 26
Gujarat                        26
Rajasthan                      25
Odisha                         21
Kerala                         21
Telangana                      17
Punjab                         15
Assam                          14
Jharkhand                      14
Chhattisgarh                   11
Haryana                        10
Delhi                           7
Jammu and Kashmir               6
Uttarakhand                     5
Himachal Pradesh                5
Meghalaya                       2
Goa                             2
Dadra and Nagar Haveli          2
Arunachal Pradesh               2
Manipur                         2
Tripura                         2
Mizoram                         1
Sikkim  

## Minister and Gender Distribution

A cross-tabulation was used to examine the distribution of representatives
by Minister status and Gender.

- 70 female and 418 male representatives were classified as non-ministers.
- 12 female and 59 male representatives were classified as ministers.
- Among female representatives, 14.6% were classified as ministers.
- Among male representatives, 12.4% were classified as ministers.

The dataset is strongly male-dominated overall, with 477 male and 82 female
representatives. However, the proportion classified as ministers is relatively
similar between the two gender groups.

This is a descriptive observation from the dataset and does not establish
that gender influences ministerial appointment.

In [14]:
pd.crosstab(df["Minister"],df["Gender"])

Gender,Female,Male
Minister,,
No,70,418
Yes,12,59


## Minister and Gender Distribution

A cross-tabulation was used to examine the distribution of representatives
by Minister status and Gender.

- 70 female and 418 male representatives were classified as non-ministers.
- 12 female and 59 male representatives were classified as ministers.
- Among female representatives, 14.6% were classified as ministers.
- Among male representatives, 12.4% were classified as ministers.

The dataset is strongly male-dominated overall, with 477 male and 82 female
representatives. However, the proportion classified as ministers is relatively
similar between the two gender groups.

This is a descriptive observation from the dataset and does not establish
that gender influences ministerial appointment.

In [15]:
pd.crosstab(df["State"],df["Gender"])

Gender,Female,Male
State,,
Andaman and Nicobar Islands,0,1
Andhra Pradesh,4,22
Arunachal Pradesh,0,2
Assam,1,13
Bihar,3,39
Chandigarh,1,0
Chhattisgarh,3,8
Dadra and Nagar Haveli,1,1
Daman and Diu,0,1


## State and Gender Distribution

A cross-tabulation was used to examine gender representation across
states and Union Territories.

- The dataset contains 477 male and 82 female representatives.
- Male representatives outnumber female representatives in most states
  and Union Territories.
- Uttar Pradesh has the largest number of female representatives (12),
  followed by West Bengal (11) and Maharashtra (8).
- Odisha has a relatively high proportion of female representatives,
  with 7 of its 21 representatives being female.
- Small states and Union Territories should be interpreted cautiously
  because their total number of representatives is very small.

Absolute numbers and percentages should be distinguished when comparing
gender representation across states.

In [16]:
df[df["Age"].isna()][
    ["Name","Gender","Party","State","Minister"]
]

,Name,Gender,Party,State,Minister
120,"Dinesh Lal ""Nirahua"" Yadav",Male,Bharatiya Janata Party,Uttar Pradesh,No
146,Ghanshyam Singh Lodhi,Male,Bharatiya Janata Party,Uttar Pradesh,No
187,Jasbir Singh,Male,Indian National Congress,Punjab,No
514,Sushil Kumar Rinku,Male,Aam Aadmi Party,Punjab,No
547,Vijayakumar alias Vijay Vasanth,Male,Indian National Congress,Tamil Nadu,No


## Missing Age Values

Age data is missing for 5 representatives.

The affected records were inspected individually and the representatives
could still be identified using their other available information.

The missing Age values are retained as missing rather than being replaced
with estimated values at this stage.

In [17]:
df["Age"].describe

<bound method NDFrame.describe of 0      53.0
1      36.0
2      86.0
3      62.0
4      60.0
       ... 
554    53.0
555    76.0
556    53.0
557    72.0
558    39.0
Name: Age, Length: 559, dtype: float64>

## Age Distribution

- Age data is available for 554 of the 559 representatives.
- 5 Age values are missing.
- The mean age is approximately 59.2 years, while the median age is 60 years.
- The middle 50% of representatives are between 51 and 68 years old.
- The youngest representative is 30 years old and the oldest is 93 years old.
- The missing Age values are retained as missing rather than being replaced
  with an estimated value at this stage.

In [18]:
df["Start of Term"] = pd.to_datetime(
    df["Start of Term"],
    format="%d-%m-%Y"
)

## Date Conversion

The `Start of Term` column was converted from string format to
datetime format using `pd.to_datetime()`.

The original dates were in DD-MM-YYYY format. Explicit date formatting
was used to ensure the dates were parsed correctly.

This conversion allows date-based operations and calculations to be
performed later in the analysis.

In [19]:
df["Start of Term"].dtype

dtype('<M8[us]')

In [20]:
df["End of Term Date"] = pd.to_datetime(
    df["End of Term"].replace("In Office", pd.NaT),
    dayfirst=True
)

## End of Term Date Conversion

A new `End of Term Date` column was created from the original
`End of Term` column.

- Recorded dates were converted to datetime format.
- Representatives listed as `In Office` were represented as `NaT`
  because they do not have an end date in the dataset.
- The original `End of Term` column was retained to preserve the
  original information.

In [21]:
df[["End of Term", "End of Term Date"]].head(10)

,End of Term,End of Term Date
0,In Office,NaT
1,In Office,NaT
2,In Office,NaT
3,In Office,NaT
4,In Office,NaT
5,In Office,NaT
6,In Office,NaT
7,In Office,NaT
8,In Office,NaT
9,In Office,NaT


In [22]:
df.dtypes

Name                                                                str
Constituency                                                        str
Minister                                                            str
Attendance                                                          str
Debates                                                           int64
                                                              ...      
Questions (Atomic Energy)                                         int64
Questions (Heavy Industries)                                      int64
Questions (Space)                                                 int64
Questions (Personnel, Public Grievances And Pensions)             int64
End of Term Date                                         datetime64[us]
Length: 97, dtype: object

## Data Preparation Complete

The initial data preparation and quality checks are now complete.

The dataset has been inspected for:

- Missing values
- Duplicate records
- Categorical distributions
- Demographic information
- Party and state representation
- Minister status
- Term dates
- Data types

The raw dataset remains unchanged, while necessary transformations
such as date conversion have been performed within the analysis
environment.

The cleaned dataset is now ready for the next stage:

**Exploratory Data Analysis (EDA).**

In [23]:
df.to_csv("../data/processed/17th_cleaned.csv", index=False)